### PR TITLE
[2969] Disable commands when text or attachments are present and disa…

### PR DESCRIPTION
**Pull request by MarinTolic**
### 🎯 Goal

Attachments should not be mixed with commands because the two are incompatible.
In order to signal that to the user we aim to remove the presence of commands suggestions and disable the commands buttom when attachments are present.

The user is also now unable to attach new content until they have dismissed the suggestions popup.

### 🛠 Implementation details

Introduced a new `SuggestionListControllerListener` to notify 'MessageInputView' when commands, mentions, and visibility changes occur.

Check for atttachments in `selectedAttachmentsCountListener` inside `MessageInputView` and enable/ disable command popups accordingly.

### 🎨 UI Changes

_Add relevant screenshots_

| Before | After |
| --- | --- |
| with_attachment_before | with_attachments_after |
| with_commands_before | with_commands_after |
| with_mentions_before | with_mentions_after |

### 🧪 Testing

This is a permutation based improvement so please follow the given instructions and feel free to improvise scenarios beyond them as it is easy to miss an edge case here

Scenario one:

1. Select an attachments or attachment 
2. Type in "/"

Expected result: commands popup should not be visible 

Scenario Two:

1. Type in "/"

Expected result: Commands popup should be visible, attachment button disabled

Scenario Three:

1. Type in "@"

Expected result: Mentions popup should be visible, attachment button disabled

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

experience

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 1]
```diff
@@ -527,6 +549,27 @@ public class MessageInputView : ConstraintLayout {
...
+                isSuggestionListUiVisible = isVisible
+                binding.attachmentsButton.isEnabled = !isSuggestionListUiVisible && !inputContainsCommands
+            }
+
+            override fun containsCommands(doesContain: Boolean) {
+                inputContainsCommands = doesContain
+                binding.attachmentsButton.isEnabled = !isSuggestionListUiVisible && !inputContainsCommands
+            }
+
+            override fun containsMentions(doesContain: Boolean) {}
```
Comment (MarinTolic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:570]: Not relevant in our case but it might be handy for users so I decided to put it in the `SuggestionListControllerListener` interface.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListControllerListener.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package io.getstream.chat.android.ui.suggestion.list
+
+/**
+ * A listener meant to be used for [SuggestionListController].
+ */
+public interface SuggestionListControllerListener {
+
+    /**
+     * Called when [SuggestionListUi] changes visibility.
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListControllerListener.kt:9]: Could we document the SuggestionListUi too?

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt [hunk 3]
```diff
@@ -54,21 +105,34 @@ public class SuggestionListController(
...
     public fun isSuggestionListVisible(): Boolean {
         return suggestionListUi.isSuggestionListVisible()
     }
 
     private suspend fun renderSuggestions(suggestions: Suggestions) = withContext(DispatcherProvider.Main) {
         suggestionListUi.renderSuggestions(suggestions)
+        suggestionListControllerListener?.onSuggestionListUiVisibilityChanged(suggestionListUi.isSuggestionListVisible())
     }
 
     private fun String.isCommandMessage() = COMMAND_PATTERN.matcher(this).find()
```
[the remark this answers is not in the document]
Comment (filbabic) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt:138]: It kind of is and isn't.

UX-wise, client side, we cannot know that this is a command, because it shouldn't be.

The reason BE parses it is probably if it splits based on "/" and then gets the entire second part as one string. But htis can be treated as a command + another word, as a non-command, or as a two word command.

I think our regex is okay and users can trick it and push it to BE but nothing should happen I think.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 4]
```diff
@@ -527,6 +549,27 @@ public class MessageInputView : ConstraintLayout {
         }
     }
 
+    /**
+     * Creates an instance of [SuggestionListControllerListener].
+     *
+     * Used to modify attachment button visibility to discourage mixing
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:555]: I would rephrase the comment here.

```
Used to disable integration buttons, based on if commands or attachments are added to the message.
```

Discouraging makes sense but at the same time implies it can be broken and ignored. We shouldn't really _encourage_ such behavior and just be straight up with what this does.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt (hunk 1 above)
Comment (filbabic) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:570]: Sounds good!

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 5]
```diff
@@ -527,6 +549,27 @@ public class MessageInputView : ConstraintLayout {
...
+     * commands and attachments since they are not compatible.
+     */
+    private fun createSuggestionsListControllerListener(): SuggestionListControllerListener =
+        object : SuggestionListControllerListener {
+            override fun onSuggestionListUiVisibilityChanged(isVisible: Boolean) {
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:560]: Could we potentially wrap this in a class that's called `DefaultSuggestionListControllerListener` and then hold the values/properties `isSuggestionListUiVisible` and `inputContainsCommands` there?

It seems like we're adding two variables to the message input view just to keep them around to disable the UI, but I'd much rather receive an emission from the listener here that's a composite of the checks we have:

E.g. 
```
          override fun onSuggestionListStateChanged(shouldEnableAttachments: Boolean) {
                binding.attachmentsButton.isEnabled = shouldEnableAttachments
            }
```

Just have a single function (or a lambda param) that emits if the attachments should be enabled or disabled. This way we don't have to calculate anything ourselves and the logic is featured in one place.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 6]
```diff
@@ -622,7 +665,17 @@ public class MessageInputView : ConstraintLayout {
 
                     refreshControlsState()
                     handleKeyStroke()
-                    messageInputDebouncer?.submitSuspendable { suggestionListController?.onNewMessageText(messageText) }
+
+                    // Debouncing when clearing the input will cause the suggestion list
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:669]: I would use the /**/ format of comments here if possible

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 7]
```diff
@@ -622,7 +665,17 @@ public class MessageInputView : ConstraintLayout {
...
+                    // popup to appear briefly after clearing the input in certain cases.
+                    if (messageText == String.EMPTY) {
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:671]: could we just say .isEmpty() or .isEmptyOrBlank()

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt [hunk 8]
```diff
@@ -12,32 +12,83 @@ import kotlinx.coroutines.withContext
...
+     * Enables or disables mention suggestions.
+     *
+     * Reactively recomputes suggestions upon setting.
+     */
     public var mentionsEnabled: Boolean by Delegates.observable(true) { _, _, _ -> computeSuggestions() }
+
+    /**
+     * Enables or disables commands suggestions.
+     *
+     * Reactively recomputes suggestions upon setting.
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt:57]: Really nice work

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt (hunk 3 above)
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/suggestion/list/SuggestionListController.kt:138]: Also slight comment here

**Review by filbabic — COMMENTED** [review]
I did leave a few comments to take a look at, but for the most part this is great! :]

Once we clean those up, I can take another look, approve and merge!

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] Bugs validated (bugfixes)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt (hunk 7 above)
Comment (MarinTolic) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:671]: Yes, certainly.

Sometimes I do these things because of code consistency. `String.EMPTY` is used elsewhere so I simply continued the chain.

I'll change the other implementations too.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt [hunk 9]
```diff
@@ -527,6 +539,16 @@ public class MessageInputView : ConstraintLayout {
...
 
+    /**
+     * Creates an instance of [SuggestionListControllerListener].
+     *
+     * Used to disable integration buttons, based on if commands or attachments are added to the message.
+     */
+    private fun createSuggestionsListControllerListener(): DefaultSuggestionListControllerListener =
+        DefaultSuggestionListControllerListener { shouldEnable ->
+            binding.attachmentsButton.isEnabled = shouldEnable
+        }
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/input/MessageInputView.kt:550]:
```suggestion
        DefaultSuggestionListControllerListener { shouldEnableAttachments ->
            binding.attachmentsButton.isEnabled = shouldEnableAttachments
        }
```

**Review by filbabic — APPROVED** [review]
